srv/src/main/java/my/bookshop/handlers/CatalogServiceHandler.java

```diff
@@ -2,56 +2,223 @@
+	}
+
+	/**
+	 * Recalculates and sets the book rating after a new review for the given book.
+	 * 
+	 * @param context {@link ReviewContext}
+	 */
+	@After(entity = Books_.CDS_NAME)
+	public void afterAddReview(AddReviewContext context) {
+		String bookId = (String) analyzer.analyze(context.getCqn()).targetKeys().get(Books.ID);
+		bookRatingService.setBookRating(bookId);
 	}
 
 	@After(event = CdsService.EVENT_READ)
 	public void discountBooks(Stream<Books> books) {
-		books.filter(b -> b.getTitle() != null).forEach(
-			b -> {
-				loadStockIfNotSet(b);
-				discountBooksWithMoreThan111Stock(b);
+		books.filter(b -> b.getTitle() != null).forEach(b -> {
+			loadStockIfNotSet(b);
+			discountBooksWithMoreThan111Stock(b);
+		});
+	}
+
+	@After
+	public void setIsReviewable(CdsReadEventContext context, List<Books> books) {
+		String user = context.getUserInfo().getName();
+
+		Result result = db
+				.run(Select.from(CatalogService_.REVIEWS).columns(Reviews_::book_ID).where(r -> r.reviewer().eq(user)));
+		Set<String> reviewedBooks = result.streamOf(Reviews.class).map(Reviews::getBookId).collect(Collectors.toSet());
+
+		for (Books book : books) {
+			if (reviewedBooks.contains(book.getId())) {
+				book.setIsReviewable(false);
 			}
-		);
+		}
+	}
+
+	@On
+	public void onSubmitOrder(SubmitOrderContext context) {
+		Integer amount = context.getAmount();
+		String bookId = context.getBook();
+
+		Optional<Books> book = db.run(Select.from(BOOKS).columns(Books_::stock).byId(bookId)).first(Books.class);
+
+		book.orElseThrow(() -> new ServiceException(ErrorStatuses.NOT_FOUND, MessageKeys.BOOK_MISSING)
+				.messageTarget(Books_.class, b -> b.ID()));
+
+		Integer stock = book.map(Books::getStock).get();
+
+		if (stock >= amount) {
+			db.run(Update.entity(BOOKS).byId(bookId).data(Books.STOCK, stock -= amount));
+
+			ReturnCatalogServiceSubmitOrder result = ReturnCatalogServiceSubmitOrder.create();
+			result.setStock(stock);
+
+			context.setResult(result);
+			context.setCompleted();
+		} else {
+			throw new ServiceException(ErrorStatuses.CONFLICT, String.format("%d exceeds stock for book", amount));
+		}
 	}
 
 	private void discountBooksWithMoreThan111Stock(Books b) {
-		if(b.getStock() > 111) {
+		if (b.getStock() != null && b.getStock() > 111) {
 			b.setTitle(String.format("%s -- 11%% discount", b.getTitle()));
 		}
 	}
 
 	private void loadStockIfNotSet(Books b) {
-		if (b.getStock() == null) {
-			b.setStock(db.run(Select.from(BOOKS).byId(b.getId()).columns(Books_::stock)).single(Books.class).getStock());
+		if (b.getId() != null && b.getStock() == null) {
+			b.setStock(
+					db.run(Select.from(BOOKS).byId(b.getId()).columns(Books_::stock)).single(Books.class).getStock());
+		}
+	}
+
+	private void validateReview(Reviews review) {
+		validateBook(review);
+		validateRating(review);
+	}
+
+	private void validateRating(Reviews review) {
+		Integer rating = review.getRating();
+		if (rating == null) {
```

Comment:
Is this a valid scenario?

Reply:
When using the Fiori UI this cannot happen, as rating will always be set. But, if sending the request via Postman for example while not setting the rating in your payload this can potentially happen.

Reply:
But should we not rather throw an exception in this case instead of generating a random rating?

Reply:
Yeah we could, I just implemented it in the same way it is in the nodejs sample.

Reply:
Ok, I wonder what the motivation was. @David-Kunz Could you comment?

Reply:
@MarcelMerkle , I didn't create this handler, but from a 'business scenario' it doesn't make any sense to generate a random rating. I guess it was just done for easy 'generation' of many ratings, which can come handy in demos.